### DISCUSSION TITLE
Concatenating lists during struct unification

### DISCUSSION THREAD

**Discussion by SanjayVas** [cue-lang/cue#1548] [Q&A] [answered]
When unifying structs, I'm trying to figure out how to get their list fields to be concatenated rather than having each list item be unified.

For example:

```
#Def: {
  l: [...{...}]
  l: [{foo: "bar"}, ...]
}

_value: #Def & {
  // How to concatenate l with [{bar: "baz"}, ...] here?
}

value: _value & {
  // How to concatenate l with [{baz: "box", ...}] here?
}
```

Desired result:
```
value: {
  l: [{foo: "bar"}, {bar: "baz"}, {baz: "box"}, ...]
}
```

**Comment by verdverm** [comment, marked as answer]
The best way to do this is to build them up as a struct (essentially a map) and use list comprehension to turn that into a list (in your definition).

Associative lists are on the todo list #14 

```cue
#Def: {
  l: [ for _,val in m { val } ]
  m: {...}
  m: foo: { foo: "bar" }
}

_value: #Def & {
  m: baz: { baz: "baz" }
}

value: _value & {
  m: box: { box: "box" }
}
```

You'll need some way to have unique keys in the map / struct

**Reply by kaos** [reply]
Will the order of the list elements be predictable and stable?

**Reply by verdverm** [reply]
no, there are no order guarantees at the moment with CUE, see #474 

They have been "stable" in practice, but that can change between versions, so not something to necessarily rely on

A more concrete example: https://github.com/hofstadter-io/hof/blob/_dev/schema/dm/dm.cue#L43

I use this to get stable output through code generation for Go struct field order and diffs of the files / models in other places
